### PR TITLE
Return values from pickles

### LINKED ISSUE #10968: Return values from pickles circuits

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR updates pickles to allows circuits to return values as part of its public input as well as consuming them.

In particular, this will allow us to construct a `Parties.t` within a circuit, instead of needing to precompute it.

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [X] Does this close issues? List them

* Closes #10968

### CODE REVIEW

File: src/lib/pickles/pickles.ml [hunk 1]
```diff
@@ -1257,6 +1354,153 @@ let%test_module "test no side-loaded" =
       assert (
         Promise.block_on_async_exn (fun () ->
             Tree_proof.Proof.verify_promise Tree_proof.example ) )
+
+    module Tree_proof_return = struct
```
Comment (mitschabaude) [inline src/lib/pickles/pickles.ml:1358]: thanks for adding these examples. they make it very clear what this PR achieves!

File: src/lib/pickles/pickles.ml [hunk 2]
```diff
@@ -1257,6 +1354,153 @@ let%test_module "test no side-loaded" =
...
+                ; account_creation_fee = Unsigned.UInt64.of_int 0
+                ; fork = None
+                }
+              ~choices:(fun ~self ->
+                [ { identifier = "main"
+                  ; prevs = [ No_recursion_return.tag; self ]
+                  ; main =
+                      (fun [ _; prev ] () ->
+                        let is_base_case =
+                          exists Boolean.typ ~request:(fun () -> Is_base_case)
```
Comment (mitschabaude) [inline src/lib/pickles/pickles.ml:1400]: 👍🏻 for using `handler` / `~request` in this example. this actually taught me what handlers are all about :D i wonder whether I should use one for snarkyjs to pass in private inputs, would enable me to remove some global variable hacks

File: src/lib/pickles/pickles.ml [hunk 3]
```diff
@@ -1257,6 +1354,153 @@ let%test_module "test no side-loaded" =
...
+                ; account_creation_fee = Unsigned.UInt64.of_int 0
+                ; fork = None
+                }
+              ~choices:(fun ~self ->
+                [ { identifier = "main"
+                  ; prevs = []
+                  ; main =
+                      (fun [] x ->
+                        dummy_constraints () ;
+                        ([], Field.(add one) x) )
```
Comment (mitschabaude) [inline src/lib/pickles/pickles.ml:1486]: combining input + output makes for a really slick API 💯

File: src/lib/pickles_types/hlist.mli [hunk 4]
```diff
@@ -842,6 +993,88 @@ module H4_2 : sig
...
+      | ( :: ) :
+          ('a1, 'a2, 'a3, 'a4, 's1, 's2, 's3, 's4) A.t
+          * ('b1, 'b2, 'b3, 'b4, 's1, 's2, 's3, 's4) t
+          -> ('a1 * 'b1, 'a2 * 'b2, 'a3 * 'b3, 'a4 * 'b4, 's1, 's2, 's3, 's4) t
+
+    val length : ('t1, 't2, 't3, 't4, 'e1, 'e2, 'e3, 'e4) t -> 't1 Length.n
+  end
+end
+
+(** Data type of heterogeneous lists whose content type varies over four type
```
Comment (mitschabaude) [inline src/lib/pickles_types/hlist.mli:1017]: nit: the comment should say "...varies over six type parameters..."

File: src/lib/pickles_types/hlist.mli [hunk 5]
```diff
@@ -842,6 +993,88 @@ module H4_2 : sig
...
+             , 'a6 * 'b6
+             , 's1
+             , 's2 )
+             t
+
+    val length : ('t1, 't2, 't3, 't4, 't5, 't6, 'e1, 'e2) t -> 't1 Length.n
+  end
+end
+
+(** Data type of heterogeneous lists whose content type varies over four type
```
Comment (mitschabaude) [inline src/lib/pickles_types/hlist.mli:1046]: nit: the comment should say "...varies over six type parameters..."

File: src/lib/pickles/inductive_rule.ml [hunk 6]
```diff
@@ -6,25 +6,58 @@ module B = struct
...
      allow predecessor proofs to conditionally fail to verify
 *)
-type ('prev_vars, 'prev_values, 'widths, 'heights, 'a_var, 'a_value) t =
+type ( 'prev_vars
+     , 'prev_values
+     , 'widths
+     , 'heights
+     , 'a_var
+     , 'a_value
+     , 'ret_var
```
Comment (deepthiskumar) [inline src/lib/pickles/inductive_rule.ml:38]: Add a bit about the return values in the comment above?

File: src/lib/pickles/pickles.ml [hunk 7]
```diff
@@ -272,36 +272,13 @@ module Prover = struct
...
-           , 'widthss
-           , 'heightss )
-           t
-end
-
-module Make (A : Statement_var_intf) (A_value : Statement_value_intf) = struct
-  module IR = Inductive_rule.T (A) (A_value)
+module Make
+    (A : Statement_var_intf)
+    (A_value : Statement_value_intf)
```
Comment (deepthiskumar) [inline src/lib/pickles/pickles.ml:277]: Rename these to `Statement_var` and `Statement_value`? or `Arg_var`/`Arg_value` to go with the function analogy. Makes it slightly easier to read the code below

File: src/lib/prover/prover.ml [hunk 8]
```diff
@@ -114,8 +114,10 @@ module Worker_state = struct
                    state_for_handler pending_coinbase =
                  let%map.Async.Deferred res =
                    Deferred.Or_error.try_with ~here:[%here] (fun () ->
-                       let t = ledger_proof_opt chain next_state t in
-                       let%map.Async.Deferred proof =
+                       let txn_snark_state, txn_snark_proof =
```
Comment (deepthiskumar) [inline src/lib/prover/prover.ml:117]: rename `txn_snark_state` to `txn_snark_statement`?

File: src/lib/pickles/reduced_me_only.ml [hunk 9]
```diff
@@ -12,7 +12,7 @@ open Backend
 module Step = struct
   [%%versioned
   module Stable = struct
-    module V1 = struct
+    module V2 = struct
```
Comment (deepthiskumar) [inline src/lib/pickles/reduced_me_only.ml:15]: revert?

**Review by deepthiskumar — APPROVED** [review]
blockchain and transaction snark changes looks good. Just some nits on naming

**Review by imeckler — APPROVED** [review]
Nice!

**Pull request merged**